### PR TITLE
[analyzer] Fix crash on dereference invalid return value of getAdjustedParameterIndex()

### LINKED ISSUE #78810: [clang][analyzer] `ExprEngineCXX` Segfault while trying to analyze valid code.
Environment: `WSL2 with Ubuntu 23.04 @ Windows 10 Pro 21H2`

Clang++ Version:
```
nyyakko@DESKTOP-7N72PNH:/mnt/c/Users/nyako/Git/clangy$ clang++ --version
Ubuntu clang version 18.0.0 (++20240119042255+5f41cef58f72-1~exp1~20240119162419.571)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

Clang-Tidy Version:
```
nyyakko@DESKTOP-7N72PNH:/mnt/c/Users/nyako/Git/clangy$ clang-tidy --version
Ubuntu LLVM version 18.0.0
  Optimized build.
```

Snippet to reproduce the crash:
```cpp
struct S
{
    constexpr auto operator==(this auto, S)
    {
        return true;
    }
};

int main()
{
    return S {} == S {};
}
```

Backtrace:

```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /usr/bin/clang-tidy --extra-arg=-Wno-unknown-warning-option --extra-arg=-Wno-ignored-optimization-argument --extra-arg=-Wno-unused-command-line-a
rgument -warnings-as-errors=* --use-color --p --extra-arg-before=--driver-mode=g++ /mnt/c/Users/nyako/Git/clangy/clangy/source/main.cpp -- /usr/bin/c++ -I/mnt/c/Users/nyako
/Git/clangy/clangy/include -I/mnt/c/Users/nyako/Git/clangy/clangy/include/clangy -g -std=gnu++23 -Werror -Wall -Wextra -Wshadow -Wnon-virtual-dtor -Wold-style-cast -Wcast-a
lign -Wunused -Woverloaded-virtual -Wpedantic -Wconversion -Wsign-conversion -Wnull-dereference -Wdouble-promotion -Wimplicit-fallthrough -MD -MT clangy/CMakeFiles/clangy.d
ir/source/main.cpp.o -MF clangy/CMakeFiles/clangy.dir/source/main.cpp.o.d -o clangy/CMakeFiles/clangy.dir/source/main.cpp.o -c /mnt/c/Users/nyako/Git/clangy/clangy/source/m
ain.cpp
1.      <eof> parser at end of file
2.      While analyzing stack: 
        #0 Calling main()
3.      /mnt/c/Users/nyako/Git/clangy/clangy/source/main.cpp:11:12: Error evaluating statement
4.      /mnt/c/Users/nyako/Git/clangy/clangy/source/main.cpp:11:12: Error evaluating statement
 #0 0x00007f12fc7bbc36 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/llvm-18/bin/../lib/libLLVM-18.so.1+0xd85c36)
 #1 0x00007f12fc7b9c60 llvm::sys::RunSignalHandlers() (/usr/lib/llvm-18/bin/../lib/libLLVM-18.so.1+0xd83c60)
 #2 0x00007f12fc7bc2fb (/usr/lib/llvm-18/bin/../lib/libLLVM-18.so.1+0xd862fb)
 #3 0x00007f12fb4e5460 (/lib/x86_64-linux-gnu/libc.so.6+0x3c460)
 #4 0x00007f1305c925ca clang::ento::ParamVarRegion::getDecl() const (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x2ca15ca)
 #5 0x00007f1305c92547 clang::ento::ParamVarRegion::getValueType() const (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x2ca1547)
 #6 0x00007f1305ccbb21 (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x2cdab21)
 #7 0x00007f1305cc6ea0 (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x2cd5ea0)
 #8 0x00007f1305ca2d97 clang::ento::ProgramState::getSVal(clang::ento::Loc, clang::QualType) const (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x2cb1d97)
 #9 0x00007f1305c7ce26 clang::ento::ExprEngine::bindReturnValue(clang::ento::CallEvent const&, clang::LocationContext const*, llvm::IntrusiveRefCntPtr<clang::ento::ProgramS
tate const>) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x2c8be26)
#10 0x00007f1305c73b58 clang::ento::ExprEngine::performTrivialCopy(clang::ento::NodeBuilder&, clang::ento::ExplodedNode*, clang::ento::CallEvent const&) (/usr/lib/llvm-18/b
in/../lib/libclang-cpp.so.18+0x2c82b58)
#11 0x00007f1305c76312 clang::ento::ExprEngine::handleConstructor(clang::Expr const*, clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&) (/usr/lib/llvm-18/bin/../li
b/libclang-cpp.so.18+0x2c85312)
#12 0x00007f1305c5a2de clang::ento::ExprEngine::Visit(clang::Stmt const*, clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&) (/usr/lib/llvm-18/bin/../lib/libclang-c
pp.so.18+0x2c692de)
#13 0x00007f1305c57bb3 clang::ento::ExprEngine::ProcessStmt(clang::Stmt const*, clang::ento::ExplodedNode*) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x2c66bb3)
#14 0x00007f1305c578df clang::ento::ExprEngine::processCFGElement(clang::CFGElement, clang::ento::ExplodedNode*, unsigned int, clang::ento::NodeBuilderContext*) (/usr/lib/l
lvm-18/bin/../lib/libclang-cpp.so.18+0x2c668df)
#15 0x00007f1305c3e9b7 clang::ento::CoreEngine::dispatchWorkItem(clang::ento::ExplodedNode*, clang::ProgramPoint, clang::ento::WorkListUnit const&) (/usr/lib/llvm-18/bin/..
/lib/libclang-cpp.so.18+0x2c4d9b7)
#16 0x00007f1305c3e521 clang::ento::CoreEngine::ExecuteWorkList(clang::LocationContext const*, unsigned int, llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>) (/us
r/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x2c4d521)
#17 0x00007f130605f5c5 (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x306e5c5)
#18 0x00007f130603f634 (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x304e634)
#19 0x00007f13059d2bac clang::MultiplexConsumer::HandleTranslationUnit(clang::ASTContext&) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x29e1bac)
#20 0x00007f1303b71496 clang::ParseAST(clang::Sema&, bool, bool) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0xb80496)
#21 0x00007f1305996b05 clang::FrontendAction::Execute() (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x29a5b05)
#22 0x00007f130590dc74 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x291cc74)
#23 0x00007f1305b85aa1 clang::tooling::FrontendActionFactory::runInvocation(std::shared_ptr<clang::CompilerInvocation>, clang::FileManager*, std::shared_ptr<clang::PCHConta
inerOperations>, clang::DiagnosticConsumer*) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x2b94aa1)
#24 0x0000556863258601 (/usr/bin/clang-tidy+0x1356601)
#25 0x00007f1305b8581f clang::tooling::ToolInvocation::runInvocation(char const*, clang::driver::Compilation*, std::shared_ptr<clang::CompilerInvocation>, std::shared_ptr<c
lang::PCHContainerOperations>) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x2b9481f)
#26 0x00007f1305b846b4 clang::tooling::ToolInvocation::run() (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x2b936b4)
#27 0x00007f1305b876e5 clang::tooling::ClangTool::run(clang::tooling::ToolAction*) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x2b966e5)
#28 0x0000556863254797 clang::tidy::runClangTidy(clang::tidy::ClangTidyContext&, clang::tooling::CompilationDatabase const&, llvm::ArrayRef<std::__cxx11::basic_string<char,
 std::char_traits<char>, std::allocator<char>>>, llvm::IntrusiveRefCntPtr<llvm::vfs::OverlayFileSystem>, bool, bool, llvm::StringRef) (/usr/bin/clang-tidy+0x1352797)
#29 0x00005568625fae3f clang::tidy::clangTidyMain(int, char const**) (/usr/bin/clang-tidy+0x6f8e3f)
#30 0x00007f12fb4cca90 (/lib/x86_64-linux-gnu/libc.so.6+0x23a90)
#31 0x00007f12fb4ccb49 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x23b49)
#32 0x00005568625f6205 _start (/usr/bin/clang-tidy+0x6f4205)
Segmentation fault
```

### PR DESCRIPTION
**Pull request by mzyKi**
fix #78810 
thanks for @Snape3058 's comment

### CODE REVIEW

**Comment by Snape3058** [comment]
Add suggested reviewers for this PR. I am unfamiliar with the engine code.

> mzyKi force-pushed the bugfix/crash-on-getAdjustedParameterIndex branch

Please do not force push to your forked repo later after receiving suggestions from other reviewers. This will make previous suggestions hard to track. We have the *squash and merge* functionality to merge all the commits on your branch into one commit in the LLVM repo when everything is done.

File: clang/test/Analysis/engine/expr-engine-cxx-crash.cpp [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// RUN: %clang_analyze_cc1 -analyzer-checker=core.DivideZero -std=c++23 -verify %s
```
Comment (steakhal) [inline clang/test/Analysis/engine/expr-engine-cxx-crash.cpp:1]:
```suggestion
// RUN: %clang_analyze_cc1 -analyzer-checker=core -std=c++23 -verify %s
```

File: clang/test/Analysis/engine/expr-engine-cxx-crash.cpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    constexpr auto operator==(this auto, S)
+    {
+        return true;
+    }
+};
+
+int main()
+{
+    return S {} == S {};
+}
```
Comment (steakhal) [inline clang/test/Analysis/engine/expr-engine-cxx-crash.cpp:15]:
```suggestion
struct S {
  bool operator==(this auto, S) {
    return true;
  }
};

int use_deducing_this() {
  return S{} == S{};
}
```

**Review by steakhal — COMMENTED** [review]
I feel like it's better than crashing, but I suspect it only resolves that, right?

File: clang/test/Analysis/engine/expr-engine-cxx-crash.cpp (hunk 2 above)
Comment (steakhal) [reply clang/test/Analysis/engine/expr-engine-cxx-crash.cpp:15]: Speaking of this test, it might be better to dedicate a file for "deducing this".

**Comment by tomasz-kaminski-sonarsource** [comment]
Thank you for creating the ticket, and fixing the issue. I am a bit concerned that the fix is hiding the symptoms and not the actual bug. After a bit of checking, I think the issue is caused by the fact that for the explicit object functions, we are creating a `CXXMemberOperatorCall` event, that performs parameter adjustment.
The corresponding code can be [found here](https://github.com/llvm/llvm-project/blob/c7fdd8c11e54585dc9d15d63de9742067e0506b9/clang/lib/StaticAnalyzer/Core/CallEvent.cpp#L1412):
```c++
 if (const auto *OpCE = dyn_cast<CXXOperatorCallExpr>(CE)) {
    const FunctionDecl *DirectCallee = OpCE->getDirectCallee();
    if (const auto *MD = dyn_cast<CXXMethodDecl>(DirectCallee))
      if (MD->isInstance())  // <-- The root of the issue
        return create<CXXMemberOperatorCall>(OpCE, State, LCtx, ElemRef);
  }
```
I think it would be fixed by replacing `MD->isInstance()` with  `isImplicitObjectMemberFunction()`. This will make use fallback to `SimpleFunctionCall`.

That would make the behavior consistent with the behavior for other explicit object functions (deducing this). 
To illustrate we have the following code, the AST and produce call event are different:
```c++
struct Foo {
  void foo();
  void bar(this Foo&);
};

void test(Foo f) {
    f.foo(); // represented as `CXXMemberCallExpr`, and creates`CXXMemberCall` event
    f.bar(); // represented as `CallExpr`, and creates `SimpleFunctionCall` ente
}
```
Live link here: https://godbolt.org/z/sjEbb6rMe.

Would you be willing to adjust PR to implement suggested change?

**Comment by mzyKi** [comment]
> That would make the behavior consistent with the behavior for other explicit object functions (deducing this).
> To illustrate we have the following code, the AST and produced CallEvent are different:

Thanks for your help very much.I only fixed the symptoms of this bug without spending much time to find the root cause. I feel  ashamed. But I am not sure whether this change has some side effect.```MD->isInstance()``` replaced with ```MD->isImplicitObjectMemberFunction()``` will miss some cases create ```<CXXMemberOperatorCall>``` in my opinion.Will these missing cases unnecessary?

**Comment by tomasz-kaminski-sonarsource** [comment]
> Thanks for your help very much.I only fixed the symptoms of this bug without spending much time to find the root cause. I feel ashamed. 

There is nothing to be ashamed of. You have spent the time to debug the crash, and localize the use of non-engaged optional, this is already helpful for the project. And addressing the crash (for example by stopping analysis) makes sure, that analysis continues and we are able to report issues from other functions. This is an improvement over the status quo, and sometimes determining the root cause is very difficult.

> But I am not sure whether this change has some side effects.`MD->isInstance()` replaced with `MD->isImplicitObjectMemberFunction()` will miss some cases create `<CXXMemberOperatorCall>` in my opinion.Are these missing cases unnecessary?

There are 4 possible choices on how an operator can be overloaded in C++:
a) free functions, including friends
b) implicit object member functions (old C++ member methods)
c) explicit object member functions (`this` parameters)
d) static member functions (you can do `operator()` and `operator[]`)

When the `CXXOperatorCallExpr` is constructed, it will always contain an object parameter on the argument list. So for the argument to parameter adjustment, we need to do:
a) nothing -> same number of params and args
b) adjust + 1 -> there is implicit object param mapped to this, and we need to expose this region
c) nothing -> object has the corresponding param, so the number is the same
d) this is a bit tricky, as `a(10)` will have 2 args and only one parameter.

Before C++23, the option didn't exist, and `isInstance()` implied that we are in case (`b`). Now `isInstance()` covers both `b` and `c`, which is leading to the crash you have found. This we need to fallback to the old status queue by using `isImplicitObjectParameter()`.

We may still not cover the case `d`, of static operators, for example, we should verify if div by zero is raised here if the operator is static, and if removing `static` helps. But I think this should be handled as a separate PR.
```C++
struct C {
  static void operator()(int x) { return 10 / x; }
};

void test(C c) {
   c(10);
}
```
If you would be interested in looking at also fixing it after this PR, I would be happy to provide my thoughts, on how to fix it.

File: clang/test/Analysis/engine/expr-engine-cxx-crash.cpp (hunk 2 above)
Comment (mzyKi) [reply clang/test/Analysis/engine/expr-engine-cxx-crash.cpp]: fixed in 5ac52a8

**Comment by mzyKi** [comment]
> > Thanks for your help very much.I only fixed the symptoms of this bug without spending much time to find the root cause. I feel ashamed.
> 
> There is nothing to be ashamed of. You have spent the time to debug the crash, and localize the use of non-engaged optional, this is already helpful for the project. And addressing the crash (for example by stopping analysis) makes sure, that analysis continues and we are able to report issues from other functions. This is an improvement over the status quo, and sometimes determining the root cause is very difficult.
> 
> > But I am not sure whether this change has some side effects.`MD->isInstance()` replaced with `MD->isImplicitObjectMemberFunction()` will miss some cases create `<CXXMemberOperatorCall>` in my opinion.Are these missing cases unnecessary?
> 
> There are 4 possible choices on how an operator can be overloaded in C++: a) free functions, including friends b) implicit object member functions (plain old C++ member functions) c) explicit object member functions (available since C++23, using `this` parameters) d) static member functions (available since C++23, for `operator()` and `operator[]`)
> 
> When the `CXXOperatorCallExpr` is constructed, it will always contain an object parameter on the argument list. So for the argument to parameter adjustment, we need to do: a) nothing -> same number of params and args b) adjust + 1 -> there is an implicit object param mapped to this, and we need to expose information about object pointed by `this` region c) nothing -> object has the corresponding param, so the number is the same d) this is a bit tricky, as `a(10)` will have 2 args and only one parameter.
> 
> Before C++23, the option didn't exist, and `isInstance()` implied that we are in case (`b`). Now `isInstance()` covers both `b` and `c`, which is leading to the crash you have found. This we need to fallback to the old status quo by using `isImplicitObjectParameter()`.
> 
> We may still not cover the case `d`, of static operators, for example, we should verify if div by zero is raised here if the operator is static, and if removing `static` helps. But I think this should be handled as a separate PR.
> 
> ```c++
> struct C {
>   static void operator()(int x) { return 10 / x; }
> };
> 
> void test(C c) {
>    c(10);
> }
> ```
> 
> If you would be interested in looking at also fixing it after this PR, I would be happy to provide my thoughts, on how I would approach it.

Thanks for your explanation and I'm willing to do the work later.I also hope that you will give me some suggestions about that.
I test the following code:
```cpp
struct C {
  static int operator()(int x) { return 10 / x; }
};

void test(C c) {
   c(0);
}
```

run command ```clang-tidy -checks=-*,clang-analyzer-core.DivideZero test.cpp -- -std=c++23``` and it show me nothing detected.
We need to fix it after?

File: clang/test/Analysis/engine/expr-engine-cxx-crash.cpp [hunk 3]
```diff
@@ -0,0 +1,11 @@
+// RUN: %clang_analyze_cc1 -analyzer-checker=core -std=c++23 -verify %s
+// expected-no-diagnostics
+
+struct S {
+  bool operator==(this auto, S) {
+    return true;
+  }
+};
+int use_deducing_this() {
+  return S{} == S{};
```
Comment (steakhal) [inline clang/test/Analysis/engine/expr-engine-cxx-crash.cpp:10]:
```suggestion
  return S{} == S{}; // no-crash
```

**Review by steakhal — COMMENTED** [review]
The patch looks a lot better this way. Thanks!

I'd recommend adding this test to the existing `clang/test/Analysis/cxx2b-deducing-this.cpp` test file to have all deducing this testcases at a common place.

**Comment by tomasz-kaminski-sonarsource** [comment]
Looking at the examples for the static operator at ([live example](https://godbolt.org/z/GhhT5ohh7)):
```c++
struct S {
  int pad;  
  static void operator()(int x, int y) {
    // clang_analyzer_dump(x);
    // clang_analyzer_dump(y); 
  }
};

void calls() {
    S s{10};
    void (*fptr)(int, int) = &S::operator();

    s(1, 2);  // A
    S::operator()(1, 2);  // B
    fptr(1, 2); // C
}
```
You can experiment with the results of the dumps to check current behavior - you need to add `debug.ExprInspection` matcher to see the output.

The calls `B` and `C` are represented as the `CallExpr` with 2 arguments, and in static analysis, they are mapped to `SimpleFunctionCall` event. This is no different from handling of any other static member function, that we have currently. 

However, in the case of `A` the AST is different, as we have `CXXOperatorCallExpr` with 3 arguments being included (the first node is called function). The additional argument is the object argument, which is not later passed to function because it is  `static`.
````
CXXOperatorCallExpr <line:13:5, col:11> 'void' '()'
|-ImplicitCastExpr <col:6, col:11> 'void (*)(int, int)' <FunctionToPointerDecay>
| `-DeclRefExpr <col:6, col:11> 'void (int, int)' lvalue CXXMethod 0xc7ecd78 'operator()' 'void (int, int)'
|-DeclRefExpr <col:5> 'S' lvalue Var 0xc7ed028 's' 'S'
|-IntegerLiteral <col:7> 'int' 1
`-IntegerLiteral <col:10> 'int' 2
````

So in this case we also need an adjustment when mapping from arguments to parameters. However, we cannot use `CXXMemberOperatorCall` as we are not calling an implicit object function, and we do not have `this` argument.
I think the best way to handle it would be to introduce a new kind of event `CXXStaticOperatorCall`.
This event should be based on `SimpleFunctionCall` (I would start it as a copy of it), with some modification derived from `CXXMemberOperatorCall`:
* constructor from `CXXOperatorCallExpr`
* `getOriginExpr` and `getOverloadedOperator` as we will always use it for `CXXOperatorCallExpr`
* `getNumArgs`, `getArgExpr`, `getAdjustedParameterIndex`, and `getASTArgumentIndex` to skip object parameters.

[ Note: We do not copy any function related to `this` object, as there is not any. ]

Then we need to add a new `CE_CXXStaticOperator` kind, it should be after `CE_Function` and before `CE_CXXMember` as this is not an instance call.
```c++
enum CallEventKind {
  CE_Function,
  CE_CXXStaticOperator,
  CE_CXXMember,
  CE_CXXMemberOperator,
  CE_CXXDestructor,
  CE_BEG_CXX_INSTANCE_CALLS = CE_CXXMember,
  CE_END_CXX_INSTANCE_CALLS = CE_CXXDestructor,
  ....
};
```

Finally, we can change `getSimpleCall` to check for static operators:
```
 if (const auto *OpCE = dyn_cast<CXXOperatorCallExpr>(CE)) {
    const FunctionDecl *DirectCallee = OpCE->getDirectCallee();
    if (const auto *MD = dyn_cast<CXXMethodDecl>(DirectCallee)) {
      if (MD->isInstance()
        return create<CXXMemberOperatorCall>(OpCE, State, LCtx, ElemRef);
      if (MD->isStatic())
         return create<CXXStaticOperatorCall>(OpCE, State, LCtx, ElemRef);
     }    
  }
```

Hope this helps. However, please open such changes in a separate PR.

File: clang/test/Analysis/engine/expr-engine-cxx-crash.cpp (hunk 3 above)
Comment (mzyKi) [reply clang/test/Analysis/engine/expr-engine-cxx-crash.cpp]: updated in [f9779b3](https://github.com/llvm/llvm-project/pull/83585/commits/f9779b3a0a818b159cd9dc54fb5fd01d5b6ccaf6)

**Review by steakhal — APPROVED** [review]

**Review by tomasz-kaminski-sonarsource — APPROVED** [review]

**Comment by steakhal** [comment]
Backporting this in PR #84194 to clang-18.

**Pull request merged**